File: solutions/139.word-break.js
/*
 * @lc app=leetcode id=139 lang=javascript
 *
 * [139] Word Break
 */

// @lc code=start
/**
 * @param {string} s
 * @param {string[]} wordDict
 * @return {boolean}
 */
var wordBreak = function (s, wordDict) {
  const dp = [true];
  for (let j = 1; j <= s.length; ++j) {
    for (let i = 0; i < j; ++i) {
      if (dp[i] && wordDict.includes(s.slice(i, j))) {
        dp[j] = true;
        break;
      }
    }
  }
  return Boolean(dp[s.length]);
};
// @lc code=end
wordBreak("leetcode", ["leet", "code"]);

File: solutions/139.word-break.rb
# @param {String} s
# @param {String[]} word_dict
# @return {Boolean}
def word_break(s, word_dict)
  dp = [true]
  (1..s.size).each do |j|
    j.downto(1) do |i|
      dp[j] = dp[i-1] && word_dict.include?(s[(i-1)...j])
      break if dp[j]
    end
  end
  dp.last
end
# @lc code=end
